protocol/proofs/anchor-state-poison-protection.md

```diff
@@ -0,0 +1,121 @@
+
+Although a basic mitigation was added in the Granite upgrade that made it possible to reset the
+anchor state to some different (valid) state, the fundamental problem still remains that poisoned
+`FaultDisputeGame` contracts may continue to resolve and further poison the anchor state registry.
+In this cascading case, the only real solution at the moment is to switch to a new game type as
+each game type has its own anchor state mapping and therefore is not impacted by any poisoned
+anchor states in other game types.
+
+Our problem statement here is therefore that we wish to be able to prevent or mitigate anchor state
+poisoning more effectively so that we can reduce the surface for issues within the fault proof
+system as a whole.
+
+## Proposed Solution
+
+We propose solving this problem by updating the `AnchorStateRegistry` to more generally be the
+source of truth for the validity of states from the perspective of the L1 proof system. All
+functionality in the `OptimismPortal` related to state validity would be moved to the
+`AnchorStateRegistry` contract. This proposed solution would simplify the `OptimismPortal`
+contract, clarify the concept of state validity from the PoV of the proof system, and solve the
+state poisoning issue within the `AnchorStateRegistry`.
+
+### Changes to the AnchorStateRegistry
+
+The `AnchorStateRegistry` would be modified to take over the blacklisting and air-gap roles that
+are currently handled by the `OptimismPortal` contract. The `AnchorStateRegistry` would expose
+several new getter functions that can be used by the `OptimismPortal` contract to determine if a
+given dispute game can be permitted to finalize a withdrawal.
+
+The `AnchorStateRegistry` would continue to keep track of the results of `FaultDisputeGame`
+contracts via the current "push" model where the `FaultDisputeGame` attempts to trigger an update
+in the `AnchorStateRegistry` on game finalization. However, instead of immediately updating the
+anchor state when a game is finalized, the `AnchorStateRegistry` would instead store the result in
+a log of game results. Game results would only be promoted to anchor states if they've passed the
+air-gap window without being blacklisted.
+
+We can avoid making significant changes to the current challenger setup by using the call to the
+`tryUpdateAnchorState` function that the `FaultDisputeGame` already makes to both store the game
+result in the game array and to use a small binary search to update the anchor state. A binary
+search is likely to be acceptable here because we can bound the starting end of the search range to
+the index of the current anchor state and only search over states that finalized after the current
+anchor state was finalized. Although this can potentially miss a few eligible anchor states under
+certain circumstances, this is an acceptable trade-off as new anchor states *will* eventually be
+found after the current anchor state.
```

Comment:
Yeah I think I like this better because it's simpler. Worst case you have a bunch of games to process, but I think we can potentially mitigate this with a few tricks like:

- First check X number of games ahead based on average publish rate, then only iterate if the game X number of games ahead is invalid.
- Add a function that allows someone to explicitly poke to update the pointer.

Reply:
I don't think I'd want to encode an expected rate of games into the contract as it could change at any time based on who happens to propose or more likely what the proposer config is.  But the typical case is that games resolve at a reasonably constant rate, so the normal case is actually that the game you promote to the new anchor state is the one immediately after the pointer. In rare cases there's some extra games published and so you want to skip a few games ahead but if we limit to checking 5 or 10 games we still cap the gas limit and it will catch up eventually.

A permissionless method that allows attempting to advance the pointer would also be useful so if it falls a long way behind we could manually poke it more frequently to get it to catch up.  Would be safe for that method to allow specifying a start for the search as well - worst case if you specify something too recent it just won't find anything to promote and will do nothing.